google/cloud/pubsub/internal/flow_controlled_publisher_tracing_connection.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsub/internal/flow_controlled_publisher_tracing_connection.h"
+#include "google/cloud/pubsub/publisher_connection.h"
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include "google/cloud/internal/opentelemetry.h"
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+namespace {
+
+/**
+ * A decorator that adds tracing for the FlowControlledPublisherConnection
+ */
+class FlowControlledPublisherTracingConnection
+    : public pubsub::PublisherConnection {
+ public:
+  explicit FlowControlledPublisherTracingConnection(
+      std::shared_ptr<pubsub::PublisherConnection> child)
+      : child_(std::move(child)) {}
+
+  ~FlowControlledPublisherTracingConnection() override = default;
+
+  future<StatusOr<std::string>> Publish(PublishParams p) override {
+    auto span = internal::MakeSpan(
+        "pubsub::FlowControlledPublisherConnection::Publish");
+    auto result = child_->Publish(std::move(p));
```

Comment:
Should this span include any information about the flow control status?

Reply:
To get the status we would need to wait for the future to be satisfied, but that would make the span not accurately represent the flow control time, so I'm letting all the status codes propagate and they will show on the parent Publish span.

Reply:
Hmmm... Maybe this should be an even in the parent span vs. a new (very short span).  Feel free to commit as-is, but if we learn anything about spans being more expensive , this might be a place to cut things.